common/scpi.c

```diff
@@ -0,0 +1,217 @@
+ */
+int
+scpi_create_message(uint8_t client, uint8_t command, uint32_t *payload,
+                    uint16_t payload_size)
+{
+	struct scpi_buffer *buffer = scpi_alloc_buffer(client);
+
+	/* Ensure there is space to put the created message. */
+	if (buffer == NULL)
+		return EAGAIN;
+
+	/* Create the message header. */
+	buffer->mem.tx_msg.command = command;
+	buffer->mem.tx_msg.sender  = 0;
+	buffer->mem.tx_msg.size    = payload_size;
+	buffer->mem.tx_msg.status  = SCPI_OK;
+
+	/* Copy the payload to the buffer, if there is one. */
+	if (payload_size > 0)
+		memcpy(&buffer->mem.tx_msg, payload, payload_size);
+
+	queue_work(scpi_send_message, buffer);
+
+	return SUCCESS;
+}
+
+/**
+ * Part 2 of SCPI message handling.
+ *
+ * Dispatch handling of the message to the appropriate function for each
+ * command. The functions for doing so are defined in a separate file to
+ * separate the API functionality from communication/synchronization code.
+ *
+ * When the command's actions are complete, and a reply message is created,
+ * enqueue a function to send the response.
+ */
+static void
+scpi_handle_message(void *param)
+{
+	struct scpi_buffer *buffer = param;
+
+	assert(buffer->busy);
+
+	/* Create a reply in tx_msg based on the contents of rx_msg. */
+	if (!scpi_handle_cmd(buffer->client, &buffer->mem.rx_msg,
+	                     &buffer->mem.tx_msg)) {
+		/* If the message does not need a reply, stop processing. */
+		buffer->busy = false;
+		return;
+	}
+
+	queue_work(scpi_send_message, buffer);
+}
+
+/**
+ * Part 1 of SCPI message handling (received messages).
+ *
+ * This is the callback from the message box framework; it is called when a new
+ * message is received. The SCPI client regains ownership of the shared memory
+ * area once the message is acknowledged, which happens when the message box
+ * driver clears the pending IRQ, immediately after this function returns.
+ * Thus, we must do our best to copy the shared memory area's contents to a
+ * safe before returning.
+ *
+ * Once the message is in a location controlled by the SCP firmware, enqueue a
+ * work item to parse, handle, and respond to it. This allows the interrupt to
+ * finish as quickly as possible.
+ */
+static void
+scpi_receive_message(struct device *dev __unused, uint8_t client, uint32_t msg)
+{
+	struct scpi_buffer *buffer;
+	struct scpi_msg    *rx_msg = &SCPI_MEM_AREA(client).rx_msg;
+
+	assert(dev == scpi_msgbox);
+	assert(client == SCPI_NONSEC_CLIENT || client == SCPI_SECURE_CLIENT);
+
+	/* Do not try to parse messages sent with a different protocol. */
+	if (msg != SCPI_VIRTUAL_CHANNEL)
+		return;
+
+	/* Ensure there is a place to put the received message. */
+	buffer = scpi_alloc_buffer(client);
+	if (buffer == NULL) {
+		error("SCPI.%u: Buffer busy, message dropped", client);
+		return;
+	}
+
+	/* Perform minimal validation of the message to optimize copying it. */
+	if (rx_msg->size > SCPI_PAYLOAD_SIZE)
```

Comment:
Without having read this section of the manual, couldn't this be problematic (i.e. cutting the payload)?

Reply:
No, because the maximum size is implementation-defined, and we advertise only allowing a payload that will fit (as part of `GET_SCP_CAP`). So if `size` is bigger than will fit, either some reserved bit is set, or the client is malicious.

Reply:
I added a comment